File: phase10/client/session.py
"""The bridge between engine state and Archipelago.

Everything here is pure: no sockets, no async, no CommonClient. That is what
makes the interesting parts -- how received items become a GameConfig, and which
location IDs a finished hand is worth -- testable without standing up a server.
"""

from __future__ import annotations

from collections import Counter
from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Any

from ..data import (
    BASE_HAND_SIZE,
    EXTRA_DRAW,
    HAND_SIZE_UPGRADE,
    HANDS_WON_MILESTONES,
    LEAN_DEAL,
    LOCATION_NAME_TO_ID,
    MAX_SKIPS,
    MULLIGAN,
    PHASE_COUNT,
    PHASE_LOCK,
    PHASE_UNLOCK,
    SCORE_REDUCTION,
    SCORE_REDUCTION_VALUE,
    SKIP_CARD,
    TIERS,
    WILD_CARD,
    WILD_THEFT,
    milestone_location_name,
    phase_location_name,
)
from ..game.cards import STOCK_WILDS
from ..game.engine import GameConfig, HandState, PhaseHand, Table
from ..game.opponents import MID, build_opponents
from ..game.game import SAVE_VERSION, Phase10Game, RoundResult

#: Traps are one-shot. Received counts only ever grow, so pending effects are
#: tracked as (received - consumed) rather than by mutating the counts.
TRAP_NAMES = (PHASE_LOCK, LEAN_DEAL, WILD_THEFT)

LEAN_DEAL_PENALTY = 2


@dataclass
class Phase10Session:
    goal: int = 0
    starting_draws: int = 4
    checks_per_phase: int = 4
    death_link: bool = False
    opponents: int = 3

    items: Counter = field(default_factory=Counter)
    consumed_traps: Counter = field(default_factory=Counter)
    mulligans_used: int = 0
    _opponent_phases: list[int] = field(default_factory=list)
    _opponent_scores: list[int] = field(default_factory=list)
    checked_locations: set[int] = field(default_factory=set)
    locked_phase: int | None = None
    last_result: RoundResult | None = None
    #: The table the current hand is being played at, opponents included.
    table: Table | None = None
    game: Phase10Game = field(default_factory=Phase10Game)

    @classmethod
    def from_slot_data(cls, slot_data: Mapping[str, Any], rng=None) -> Phase10Session:
        return cls(
            goal=int(slot_data.get("goal", 0)),
            starting_draws=int(slot_data.get("starting_draws", 4)),
            checks_per_phase=int(slot_data.get("checks_per_phase", 4)),
            death_link=bool(slot_data.get("death_link", False)),
            opponents=int(slot_data.get("opponents", 3)),
            game=Phase10Game(rng),
        )

    # The game owns the running state; the session keeps one source of truth
    # rather than a second tally that could drift from the scorecard.
    @property
    def hand(self) -> PhaseHand | None:
        return self.game.hand

    @property
    def hands_won(self) -> int:
        return self.game.rounds_won

    @property
    def cleared_phases(self) -> set[int]:
        return self.game.cleared_phases

    @property
    def score_reduction(self) -> int:
        """Points Score Reduction items have taken off the running total."""
        return self.items[SCORE_REDUCTION] * SCORE_REDUCTION_VALUE

    @property
    def total_score(self) -> int:
        """The score as the player is judged on it: raw, less reductions.

        Floored at zero. The scorecard still shows what each round actually
        cost -- a reduction forgives points, it does not rewrite history.
        """
        return max(0, self.game.total_score - self.score_reduction)

    # -- items -------------------------------------------------------------
    def set_items(self, item_names: list[str]) -> None:
        """Replace the received-item tally. AP resends the full list, so this
        is idempotent rather than incremental."""
        self.items = Counter(item_names)

    def pending(self, trap: str) -> int:
        return max(0, self.items[trap] - self.consumed_traps[trap])

    @property
    def unlocked_phases(self) -> set[int]:
        return {p for p in range(1, PHASE_COUNT + 1) if self.items[PHASE_UNLOCK.format(p)]}

    # -- configuration -----------------------------------------------------
    @property
    def config(self) -> GameConfig:
        """Build the engine's knobs from what Archipelago has handed over."""
        hand_size = BASE_HAND_SIZE + self.items[HAND_SIZE_UPGRADE]
        wilds = min(self.items[WILD_CARD], STOCK_WILDS)
        draws = self.starting_draws + self.items[EXTRA_DRAW]

        if self.pending(LEAN_DEAL):
            hand_size -= LEAN_DEAL_PENALTY
        if self.pending(WILD_THEFT):
            wilds -= 1

        return GameConfig(
            hand_size=max(4, hand_size),
            wilds_in_deck=max(0, min(wilds, STOCK_WILDS)),
            max_draws=max(1, draws),
            starting_skips=min(self.items[SKIP_CARD], MAX_SKIPS),
        )

    # -- playing -----------------------------------------------------------
    def can_play(self, phase: int) -> str | None:
        """Returns None if the phase is playable, else why not."""
        if phase not in self.unlocked_phases:
            return f"Phase {phase} is not unlocked yet."
        if self.locked_phase is not None and phase != self.locked_phase:
            return f"A Phase Lock trap is forcing you to replay Phase {self.locked_phase}."
        return None

    # -- the table ---------------------------------------------------------
    @property
    def opponent_phases(self) -> list[int]:
        """Which phase each seat is on. Grows as they clear their own."""
        if not self._opponent_phases:
            self._opponent_phases = [1] * self.opponents
        return self._opponent_phases

    @property
    def opponent_scores(self) -> list[int]:
        """Each seat's running total, the way the pad on the table works.

        Kept on the session rather than on the seat because the seats are
        rebuilt from scratch every round, exactly like `opponent_phases`.
        """
        if not self._opponent_scores:
            self._opponent_scores = [0] * self.opponents
        return self._opponent_scores

    @property
    def seats(self) -> list:
        return self.table.seats if self.table else []

    def advance_opponents(self) -> list[str]:
        """Move every seat that cleared its phase on to the next one."""
        moved = []
        for index, seat in enumerate(self.seats):
            if seat.laid_down and index < len(self.opponent_phases):
                self._opponent_phases[index] = min(PHASE_COUNT, seat.phase + 1)
                moved.append(seat.name)
        return moved

    def tally_opponents(self) -> None:
        """Score every seat on what it is caught holding.

        Must run before the round is finished, while the seats still hold the
        hands they ended with. Lower is better here as it is for you: the seat
        that went out scores nothing, and the one still sitting on a Wild pays
        twenty-five for it.
        """
        scores = self.opponent_scores
        for index, seat in enumerate(self.seats):
            if index < len(scores):
                scores[index] += seat.score

    # -- mulligans ---------------------------------------------------------
    @property
    def mulligans_left(self) -> int:
        return max(0, self.items[MULLIGAN] - self.mulligans_used)

    def can_mulligan(self) -> str | None:
        """Returns None if a Mulligan is usable right now, else why not."""
        if not self.mulligans_left:
            return "No Mulligans left."
        hand = self.hand
        if hand is None or hand.state is not HandState.IN_PROGRESS:
            return "No hand in progress."
        if hand.draws_used or hand.drew_this_turn:
            return "A Mulligan only works before your first draw."
        if hand.dig_pending:
            return "Finish the dig first."
        if hand.skips_played:
            return "A Mulligan only works before you play a Skip."
        return None

    def use_mulligan(self) -> PhaseHand:
        """Spend a Mulligan on the current hand and return it, redealt."""
        refusal = self.can_mulligan()
        if refusal:
            raise ValueError(refusal)
        hand = self.hand
        hand.redeal()
        self.mulligans_used += 1
        return hand

    def start_hand(self, phase: int) -> PhaseHand:
        refusal = self.can_play(phase)
        if refusal:
            raise ValueError(refusal)

        config = self.config
        # Consume the one-shot traps that shaped this hand.
        for trap in (LEAN_DEAL, WILD_THEFT):
            if self.pending(trap):
                self.consumed_traps[trap] += 1

        self.table = Table()
        if self.opponents:
            # Each seat carries its own phase between rounds, so the table
            # gets harder to beat as the run goes on rather than resetting
            # to three players on phase 1 every time.
            self.table.seats = build_opponents(
                self.opponents, list(self.opponent_phases),
                config, self.game.rng, MID,
            )
        return self.game.start_round(phase, config, table=self.table)

    def kill_hand(self) -> PhaseHand | None:
        """Fail the hand in progress, if there is one.

        A card game has nothing to kill, so a DeathLink death is a lost hand.
        Between rounds there is nothing to lose and an incoming death passes
        harmlessly -- returning None says so, rather than inventing a penalty
        the player cannot see coming.
        """
        hand = self.hand
        if hand is None or hand.state is not HandState.IN_PROGRESS:
            return None
        hand.mark_failed("death_link")
        return hand

    def earned_tiers(self, hand: PhaseHand) -> list[str]:
        """Which check tiers a finished hand is worth."""
        if hand.state not in (HandState.PHASE_LAID, HandState.WENT_OUT):
            return []

        # Measured at the lay-down, not at the end of the round: the round
        # carries on afterwards so the rest of the hand can be shed, and
        # counting those draws would make Under Par unearnable.
        spent = hand.draws_at_lay_down
        if spent is None:
            spent = hand.draws_used

        earned = {"Cleared"}
        if hand.state is HandState.WENT_OUT:
            earned.add("Went Out")
        if hand.used_wilds_in_layout == 0:
            earned.add("No Wilds")
        if spent <= max(1, hand.config.max_draws // 2):
            earned.add("Under Par")

        # Walk TIERS rather than the order they were collected in, so the
        # result follows the table and a reorder reaches here for free.
        # `checks_per_phase` takes a prefix: tiers past it have no location on
        # the server, so reporting one would check an ID that does not exist.
        return [tier for tier in TIERS[: self.checks_per_phase] if tier in earned]

    def finish_hand(self, hand: PhaseHand) -> list[int]:
        """Settle a finished hand. Returns newly checked location IDs."""
        tiers = self.earned_tiers(hand)
        # Both read the seats as they ended the round, so they come before
        # the round is finished and the table is torn down.
        self.tally_opponents()
        self.advance_opponents()
        result = self.game.finish_round(hand)
        self.last_result = result

        names: list[str] = []
        if result.cleared:
            self.locked_phase = None
            names += [phase_location_name(result.phase, tier) for tier in tiers]
            names += [
                milestone_location_name(n)
                for n in HANDS_WON_MILESTONES
                if n <= self.hands_won
            ]
        elif self.pending(PHASE_LOCK):
            # A failed hand under a Phase Lock pins you to this phase.
            self.consumed_traps[PHASE_LOCK] += 1
            self.locked_phase = result.phase

        new = [
            LOCATION_NAME_TO_ID[name]
            for name in names
            if LOCATION_NAME_TO_ID[name] not in self.checked_locations
        ]
        self.checked_locations.update(new)
        return new

    # -- persistence -------------------------------------------------------
    def to_payload(self) -> dict:
        """Everything the server does not already know.

        Checked locations are deliberately left out: the server is the
        authority on those, and writing our own copy back would only create
        something that could disagree with it.
        """
        return {
            "version": SAVE_VERSION,
            "game": self.game.to_payload(),
            "consumed_traps": {k: int(v) for k, v in self.consumed_traps.items() if v},
            "mulligans_used": self.mulligans_used,
            "opponent_phases": list(self._opponent_phases),
            "opponent_scores": list(self._opponent_scores),
            "locked_phase": self.locked_phase,
        }

    def load_payload(self, payload: object) -> bool:
        """Restore from a saved payload. Returns whether it took.

        Treated as untrusted input -- it arrives over the network and a partial
        restore would be worse than none.
        """
        if not isinstance(payload, dict) or payload.get("version") != SAVE_VERSION:
            return False
        if not self.game.load_payload(payload.get("game")):
            return False

        traps = payload.get("consumed_traps")
        if isinstance(traps, dict):
            try:
                self.consumed_traps = Counter(
                    {str(k): int(v) for k, v in traps.items()}
                )
            except (TypeError, ValueError):
                self.consumed_traps = Counter()

        # Absent in saves written before Mulligans did anything, so a missing
        # key restores as zero rather than refusing the whole payload.
        used = payload.get("mulligans_used")
        self.mulligans_used = used if isinstance(used, int) and used >= 0 else 0

        phases = payload.get("opponent_phases")
        if isinstance(phases, list) and all(
            isinstance(v, int) and 1 <= v <= PHASE_COUNT for v in phases
        ):
            self._opponent_phases = list(phases)

        # Absent in saves written before the seats kept score, so a missing
        # key restores as zero rather than refusing the whole payload.
        scores = payload.get("opponent_scores")
        if isinstance(scores, list) and all(
            isinstance(v, int) and v >= 0 for v in scores
        ):
            self._opponent_scores = list(scores)

        locked = payload.get("locked_phase")
        self.locked_phase = (
            locked if isinstance(locked, int) and 1 <= locked <= PHASE_COUNT else None
        )
        return True

    # -- goal --------------------------------------------------------------
    @property
    def goal_met(self) -> bool:
        if self.goal == 1:  # phase_ten
            return 10 in self.cleared_phases
        return len(self.cleared_phases) == 10

File: docs/src/session.js
// The bridge between engine state and Archipelago.
//
// Port of phase10/client/session.py. Pure: no sockets, no DOM. That is what
// makes the interesting parts testable -- how received items become a config,
// and which location IDs a finished hand is worth.
//
// Two Python behaviours easy to lose in a port, both pinned by session_test:
//   * Counter[name] is 0 for absent keys, never undefined, so arithmetic on a
//     missing item must not produce NaN.
//   * earnedTiers is computed BEFORE finishRound, because finishing the round
//     takes the hand off the game.

import {
  BASE_HAND_SIZE, EXTRA_DRAW, HANDS_WON_MILESTONES, HAND_SIZE_UPGRADE, LEAN_DEAL,
  AP_POINT, LOCATION_NAME_TO_ID, MAX_SKIPS, MAX_STORE_SLOTS, MULLIGAN,
  PHASE_COUNT, PHASE_LOCK,
  SCORE_REDUCTION,
  SCORE_REDUCTION_VALUE, SKIP_CARD, TIERS, WILD_CARD,
  WILD_THEFT, milestoneLocationName, phaseLocationName, phaseUnlock,
  storeGate, storeLocationName, storePrices,
} from "./data.js";
import { STOCK_WILDS } from "./cards.js";
import { HAND_STATE, Table, gameConfig } from "./engine.js";
import { MID, buildOpponents } from "./opponents.js";
import { Phase10Game, SAVE_VERSION, roundCleared } from "./game.js";

export const LEAN_DEAL_PENALTY = 2;

export class Phase10Session {
  constructor(opts = {}) {
    this.goal = opts.goal ?? 0;
    this.startingDraws = opts.startingDraws ?? 4;
    this.checksPerPhase = opts.checksPerPhase ?? 4;
    this.storeSlots = opts.storeSlots ?? 0;
    this.deathLink = opts.deathLink ?? false;
    this.opponents = opts.opponents ?? 3;

    this.items = new Map();
    this.consumedTraps = new Map();
    this.mulligansUsed = 0;
    //: Which store slots have been bought, so what has been spent is derived
    //: rather than stored twice and left to disagree with itself.
    this.boughtSlots = new Set();
    this._opponentPhases = [];
    this._opponentScores = [];
    //: The table the current hand is being played at, opponents included.
    this.table = null;
    this.checkedLocations = new Set();
    this.lockedPhase = null;
    this.lastResult = null;
    this.game = opts.game ?? new Phase10Game();
  }

  static fromSlotData(slotData = {}, game = null) {
    return new Phase10Session({
      goal: Number(slotData.goal ?? 0),
      startingDraws: Number(slotData.starting_draws ?? 4),
      checksPerPhase: Number(slotData.checks_per_phase ?? 4),
      storeSlots: Number(slotData.store_slots ?? 0),
      deathLink: Boolean(slotData.death_link ?? false),
      opponents: Number(slotData.opponents ?? 3),
      game: game ?? new Phase10Game(),
    });
  }

  // The game owns the running state; the session keeps one source of truth
  // rather than a second tally that could drift from the scorecard.
  get hand() {
    return this.game.hand;
  }

  get handsWon() {
    return this.game.roundsWon;
  }

  get clearedPhases() {
    return this.game.clearedPhases;
  }

  /** Points Score Reduction items have taken off the running total. */
  get scoreReduction() {
    return this.count(SCORE_REDUCTION) * SCORE_REDUCTION_VALUE;
  }

  /**
   * The score as the player is judged on it: raw, less reductions, floored at
   * zero. The scorecard still shows what each round actually cost -- a
   * reduction forgives points, it does not rewrite history.
   */
  get totalScore() {
    return Math.max(0, this.game.totalScore - this.scoreReduction);
  }

  // -- items ---------------------------------------------------------------
  /** Counter semantics: absent means zero, never undefined. */
  count(name) {
    return this.items.get(name) ?? 0;
  }

  /**
   * Replace the received-item tally. Archipelago resends the full list, so
   * this is idempotent rather than incremental, which also makes it safe to
   * call again on reconnect.
   */
  setItems(itemNames) {
    this.items = new Map();
    for (const name of itemNames) {
      this.items.set(name, this.count(name) + 1);
    }
  }

  pending(trap) {
    return Math.max(0, this.count(trap) - (this.consumedTraps.get(trap) ?? 0));
  }

  get unlockedPhases() {
    const open = new Set();
    for (let p = 1; p <= PHASE_COUNT; p += 1) {
      if (this.count(phaseUnlock(p)) > 0) open.add(p);
    }
    return open;
  }

  // -- configuration -------------------------------------------------------
  get config() {
    let handSize = BASE_HAND_SIZE + this.count(HAND_SIZE_UPGRADE);
    let wilds = Math.min(this.count(WILD_CARD), STOCK_WILDS);
    const draws = this.startingDraws + this.count(EXTRA_DRAW);

    if (this.pending(LEAN_DEAL)) handSize -= LEAN_DEAL_PENALTY;
    if (this.pending(WILD_THEFT)) wilds -= 1;

    return gameConfig({
      handSize: Math.max(4, handSize),
      wildsInDeck: Math.max(0, Math.min(wilds, STOCK_WILDS)),
      // Zero starting draws means no budget at all, which only free play
      // asks for: the Archipelago option starts at 2.
      maxDraws: this.startingDraws <= 0 ? 0 : Math.max(1, draws),
      startingSkips: Math.min(this.count(SKIP_CARD), MAX_SKIPS),
    });
  }

  // -- playing -------------------------------------------------------------
  /** Returns null if the phase is playable, else why not. */
  canPlay(phase) {
    if (!this.unlockedPhases.has(phase)) {
      return `Phase ${phase} is not unlocked yet.`;
    }
    if (this.lockedPhase !== null && phase !== this.lockedPhase) {
      return `A Phase Lock trap is forcing you to replay Phase ${this.lockedPhase}.`;
    }
    return null;
  }

  // -- the store -----------------------------------------------------------
  // Points arrive as items and buy a check outright. The seed priced each slot
  // at generation, and the gate on a slot is the sum of the cheapest prices up
  // to it -- so holding enough to reach a slot's gate means you could have
  // bought the cheaper ones instead, and any order is legal.

  /** Points received. Never goes down; spending is tracked separately. */
  get points() {
    return this.count(AP_POINT);
  }

  get pointsSpent() {
    const prices = storePrices(this.storeSlots);
    let spent = 0;
    for (const slot of this.boughtSlots) {
      if (slot >= 1 && slot <= this.storeSlots) spent += prices[slot - 1];
    }
    return spent;
  }

  get pointsLeft() {
    return Math.max(0, this.points - this.pointsSpent);
  }

  storePrice(slot) {
    return storePrices(this.storeSlots)[slot - 1];
  }

  /** Returns null if the slot is buyable right now, else why not. */
  canBuy(slot) {
    if (!this.storeSlots) return "This seed has no store.";
    if (!(slot >= 1 && slot <= this.storeSlots)) {
      return `The store has slots 1 to ${this.storeSlots}.`;
    }
    if (this.boughtSlots.has(slot)) return `Slot ${slot} is already bought.`;
    const gate = storeGate(slot);
    if (this.points < gate) {
      // The gate is on points received, not points left: it is what the seed's
      // logic was built on, so checking it here is what keeps the client from
      // reporting a location the server thinks is unreachable.
      return `Slot ${slot} opens at ${gate} points received; you have ${this.points}.`;
    }
    const price = this.storePrice(slot);
    // Unreachable while the prices ascend: any set of slots whose gates you
    // have met costs at most the largest of those gates, which you have. Kept
    // because it is what would catch a ladder that stopped ascending, and the
    // tests pin the invariant rather than this branch.
    if (this.pointsLeft < price) {
      return `Slot ${slot} costs ${price}; you have ${this.pointsLeft} unspent.`;
    }
    return null;
  }

  /** Buy a slot. Returns the location ID to check. */
  buySlot(slot) {
    const refusal = this.canBuy(slot);
    if (refusal) throw new Error(refusal);
    this.boughtSlots.add(slot);
    const id = LOCATION_NAME_TO_ID[storeLocationName(slot)];
    this.checkedLocations.add(id);
    return id;
  }

  // -- mulligans -----------------------------------------------------------
  get mulligansLeft() {
    return Math.max(0, this.count(MULLIGAN) - this.mulligansUsed);
  }

  /** Returns null if a Mulligan is usable right now, else why not. */
  canMulligan() {
    if (!this.mulligansLeft) return "No Mulligans left.";
    const hand = this.hand;
    if (!hand || hand.state !== HAND_STATE.IN_PROGRESS) return "No hand in progress.";
    if (hand.drawsUsed || hand.drewThisTurn) {
      return "A Mulligan only works before your first draw.";
    }
    if (hand.digPending) return "Finish the dig first.";
    if (hand.skipsPlayed) return "A Mulligan only works before you play a Skip.";
    return null;
  }

  /** Spend a Mulligan on the current hand and return it, redealt. */
  useMulligan() {
    const refusal = this.canMulligan();
    if (refusal) throw new Error(refusal);
    const hand = this.hand;
    hand.redeal();
    this.mulligansUsed += 1;
    return hand;
  }

  // -- the table -----------------------------------------------------------
  /** Which phase each seat is on. Grows as they clear their own. */
  get opponentPhases() {
    if (!this._opponentPhases.length) {
      this._opponentPhases = new Array(this.opponents).fill(1);
    }
    return this._opponentPhases;
  }

  /**
   * Each seat's running total, the way the pad on the table works.
   *
   * Kept on the session rather than on the seat because the seats are rebuilt
   * from scratch every round, exactly like opponentPhases.
   */
  get opponentScores() {
    if (!this._opponentScores.length) {
      this._opponentScores = new Array(this.opponents).fill(0);
    }
    return this._opponentScores;
  }

  get seats() {
    return this.table ? this.table.seats : [];
  }

  /** Move every seat that cleared its phase on to the next one. */
  advanceOpponents() {
    const moved = [];
    this.seats.forEach((seat, index) => {
      if (seat.laidDown && index < this.opponentPhases.length) {
        this._opponentPhases[index] = Math.min(PHASE_COUNT, seat.phase + 1);
        moved.push(seat.name);
      }
    });
    return moved;
  }

  /**
   * Score every seat on what it is caught holding.
   *
   * Must run before the round is finished, while the seats still hold the
   * hands they ended with. Lower is better here as it is for you: the seat
   * that went out scores nothing, and the one still sitting on a Wild pays
   * twenty-five for it.
   */
  tallyOpponents() {
    const scores = this.opponentScores;
    this.seats.forEach((seat, index) => {
      if (index < scores.length) scores[index] += seat.score;
    });
  }

  startHand(phase, opts = {}) {
    const refusal = this.canPlay(phase);
    if (refusal) throw new Error(refusal);

    const config = this.config;
    // Consume the one-shot traps that shaped this hand.
    for (const trap of [LEAN_DEAL, WILD_THEFT]) {
      if (this.pending(trap)) {
        this.consumedTraps.set(trap, (this.consumedTraps.get(trap) ?? 0) + 1);
      }
    }
    this.table = new Table();
    if (this.opponents) {
      // Each seat carries its own phase between rounds, so the table gets
      // harder to beat as the run goes on rather than resetting to three
      // players on phase 1 every time.
      this.table.seats = buildOpponents(
        this.opponents, [...this.opponentPhases], config, this.game.random, MID,
      );
    }
    return this.game.startRound(phase, config, { table: this.table, ...opts });
  }

  /**
   * Fail the hand in progress, if there is one.
   *
   * A card game has nothing to kill, so a DeathLink death is a lost hand.
   * Between rounds there is nothing to lose and an incoming death passes
   * harmlessly -- returning null says so, rather than inventing a penalty the
   * player cannot see coming.
   */
  killHand() {
    const hand = this.hand;
    if (!hand || hand.state !== HAND_STATE.IN_PROGRESS) return null;
    hand.markFailed("death_link");
    return hand;
  }

  /** Which check tiers a finished hand is worth. */
  earnedTiers(hand) {
    if (hand.state !== HAND_STATE.PHASE_LAID && hand.state !== HAND_STATE.WENT_OUT) {
      return [];
    }

    // Measured at the lay-down, not at the end of the round: the round
    // carries on afterwards so the rest of the hand can be shed, and counting
    // those draws would make Under Par unearnable.
    const spent = hand.drawsAtLayDown ?? hand.drawsUsed;

    const earned = new Set(["Cleared"]);
    if (hand.state === HAND_STATE.WENT_OUT) earned.add("Went Out");
    if (hand.usedWildsInLayout === 0) earned.add("No Wilds");
    if (spent <= Math.max(1, Math.floor(hand.config.maxDraws / 2))) {
      earned.add("Under Par");
    }

    // Walk TIERS rather than the order they were collected in, so the result
    // follows the table and a reorder reaches here for free. checksPerPhase
    // takes a prefix: tiers past it have no location on the server, so
    // reporting one would check an ID that does not exist.
    return TIERS.slice(0, this.checksPerPhase).filter((tier) => earned.has(tier));
  }

  /** Settle a finished hand. Returns newly checked location IDs. */
  finishHand(hand) {
    // Computed first: finishing the round takes the hand off the game.
    const tiers = this.earnedTiers(hand);
    // Both read the seats as they ended the round, so they come before the
    // round is finished and the table is torn down.
    this.tallyOpponents();
    this.advanceOpponents();
    const result = this.game.finishRound(hand);
    this.lastResult = result;

    const names = [];
    if (roundCleared(result)) {
      this.lockedPhase = null;
      for (const tier of tiers) names.push(phaseLocationName(result.phase, tier));
      for (const n of HANDS_WON_MILESTONES) {
        if (n <= this.handsWon) names.push(milestoneLocationName(n));
      }
    } else if (this.pending(PHASE_LOCK)) {
      // A failed hand under a Phase Lock pins you to this phase.
      this.consumedTraps.set(PHASE_LOCK, (this.consumedTraps.get(PHASE_LOCK) ?? 0) + 1);
      this.lockedPhase = result.phase;
    }

    const fresh = [];
    for (const name of names) {
      const id = LOCATION_NAME_TO_ID[name];
      if (!this.checkedLocations.has(id)) {
        this.checkedLocations.add(id);
        fresh.push(id);
      }
    }
    return fresh;
  }

  // -- persistence ---------------------------------------------------------
  /**
   * Everything the server does not already know. Checked locations are
   * deliberately left out: the server is the authority on those, and a second
   * copy could only ever disagree with it.
   */
  toPayload() {
    const traps = {};
    for (const [name, count] of this.consumedTraps) {
      if (count) traps[name] = count;
    }
    return {
      version: SAVE_VERSION,
      game: this.game.toPayload(),
      consumed_traps: traps,
      mulligans_used: this.mulligansUsed,
      opponent_phases: [...this._opponentPhases],
      opponent_scores: [...this._opponentScores],
      bought_slots: [...this.boughtSlots].sort((a, b) => a - b),
      locked_phase: this.lockedPhase,
    };
  }

  /**
   * Restore from a saved payload. Returns whether it took. Treated as
   * untrusted input -- it arrives over the network, and a partial restore
   * would be worse than none.
   */
  loadPayload(payload) {
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) return false;
    if (payload.version !== SAVE_VERSION) return false;
    if (!this.game.loadPayload(payload.game)) return false;

    const traps = payload.consumed_traps;
    this.consumedTraps = new Map();
    if (traps && typeof traps === "object" && !Array.isArray(traps)) {
      for (const [name, count] of Object.entries(traps)) {
        if (Number.isInteger(count) && count > 0) this.consumedTraps.set(name, count);
      }
    }

    // Absent in saves written before Mulligans did anything, so a missing key
    // restores as zero rather than refusing the whole payload.
    const used = payload.mulligans_used;
    this.mulligansUsed = Number.isInteger(used) && used >= 0 ? used : 0;

    // Without this a reconnect reseats everyone on phase 1, which quietly
    // hands the player an easier table than they had earned.
    const phases = payload.opponent_phases;
    if (Array.isArray(phases)
        && phases.every((v) => Number.isInteger(v) && v >= 1 && v <= PHASE_COUNT)) {
      this._opponentPhases = [...phases];
    }

    // Absent in saves written before the seats kept score, so a missing key
    // restores as zero rather than refusing the whole payload.
    const scores = payload.opponent_scores;
    if (Array.isArray(scores)
        && scores.every((v) => Number.isInteger(v) && v >= 0)) {
      this._opponentScores = [...scores];
    }

    // Absent in saves written before the store existed, so a missing key
    // restores as nothing bought rather than refusing the whole payload.
    const bought = payload.bought_slots;
    if (Array.isArray(bought)
        && bought.every((v) => Number.isInteger(v) && v >= 1 && v <= MAX_STORE_SLOTS)) {
      this.boughtSlots = new Set(bought);
    }

    const locked = payload.locked_phase;
    this.lockedPhase =
      Number.isInteger(locked) && locked >= 1 && locked <= PHASE_COUNT ? locked : null;
    return true;
  }

  // -- goal ----------------------------------------------------------------
  get goalMet() {
    if (this.goal === 1) return this.clearedPhases.has(10);
    return this.clearedPhases.size === 10;
  }
}
